File: src/java/magmac/app/io/SafeFiles.java
package magmac.app.io;

import magmac.api.None;
import magmac.api.Option;
import magmac.api.Some;
import magmac.api.collect.list.JVMList;
import magmac.api.iter.Iter;
import magmac.api.result.Err;
import magmac.api.result.Ok;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class SafeFiles {
    public static Option<IOException> writeString(Path target, String output) {
        try {
            Files.writeString(target, output);
            return new None<>();
        } catch (IOException e) {
            return new Some<>(e);
        }
    }

    public static IOResult<Iter<Path>> walk(Path sourceDirectory) {
        try {
            var stream = Files.walk(sourceDirectory);
            return new InlineIOResult<>(new Ok<>(new JVMList<>(stream.collect(Collectors.toList())).iter()));
        } catch (IOException e) {
            return new InlineIOResult<>(new Err<>(e));
        }
    }

    public static IOResult<String> readString(Path source) {
        try {
            return new InlineIOResult<>(new Ok<>(Files.readString(source)));
        } catch (IOException e) {
            return new InlineIOResult<>(new Err<>(e));
        }
    }

    public static Option<IOException> createDirectories(Path targetParent) {
        try {
            Files.createDirectories(targetParent);
            return new None<>();
        } catch (IOException e) {
            return new Some<>(e);
        }
    }
}

File: src/web/magmac/app/io/SafeFiles.ts
import { None } from "../../../magmac/api/None";
import { Option } from "../../../magmac/api/Option";
import { Some } from "../../../magmac/api/Some";
import { JVMList } from "../../../magmac/api/collect/list/JVMList";
import { Iter } from "../../../magmac/api/iter/Iter";
import { Err } from "../../../magmac/api/result/Err";
import { Ok } from "../../../magmac/api/result/Ok";
import { IOException } from "../../../java/io/IOException";
import { Files } from "../../../java/nio/file/Files";
import { Path } from "../../../java/nio/file/Path";
import { Collectors } from "../../../java/util/stream/Collectors";
import { Stream } from "../../../java/util/stream/Stream";
export class SafeFiles {
	public static writeString( target : Path,  output : String) : Option<IOException> {if(true){ Files.writeString( target, output);return new None<>( );;}if(true){ return new Some<>( e);;};}
	public static walk( sourceDirectory : Path) : IOResult<Iter<Path>> {if(true){  let stream : var=Files.walk( sourceDirectory);return new InlineIOResult<>( new Ok<>( new JVMList<>( stream.collect( Collectors.toList( ))).iter( )));;}if(true){ return new InlineIOResult<>( new Err<>( e));;};}
	public static readString( source : Path) : IOResult<String> {if(true){ return new InlineIOResult<>( new Ok<>( Files.readString( source)));;}if(true){ return new InlineIOResult<>( new Err<>( e));;};}
	public static createDirectories( targetParent : Path) : Option<IOException> {if(true){ Files.createDirectories( targetParent);return new None<>( );;}if(true){ return new Some<>( e);;};}
}
